fv3config/_asset_list.py

```diff
@@ -0,0 +1,245 @@
+                                             basename,
+                                             target_location=target_location))
+    return asset_list
+
+
+def write_asset(asset, target_directory):
+    """Write file represented by asset to target_directory
+
+    Args:
+        asset (dict): an asset dict
+        target_directory (str): path to a directory in which all files will be written
+    """
+    check_asset_valid(asset)
+    source_path = os.path.join(asset['source_location'], asset['source_name'])
+    target_path = os.path.join(target_directory, asset['target_location'], asset['target_name'])
+    copy_method = asset['copy_method']
+    if not os.path.exists(os.path.dirname(target_path)):
+        os.makedirs(os.path.dirname(target_path))
+    if copy_method == 'copy':
+        copy_file(source_path, target_path)
+    elif copy_method == 'link':
+        link_file(source_path, target_path)
+    else:
+        raise ConfigError(
+            f'Behavior of copy_method {copy_method} not defined for {source_path} asset'
+        )
+
+
+def write_asset_list(asset_list, target_directory):
+    """Loop over list of assets and write them all"""
+    for asset in asset_list:
+        write_asset(asset, target_directory)
+
+
+def check_asset_valid(asset):
+    """Check asset has all of its required keys"""
+    required_asset_properties = ['source_location', 'source_name', 'target_location',
+                                 'target_name', 'copy_method']
+    for required_asset_property in required_asset_properties:
+        if required_asset_property not in asset:
+            raise ConfigError(f'Assets must have a {required_asset_property}')
+
+
+def config_to_asset_list(config):
+    """Convert a configuration dictionary to an asset list. The asset list
+    will contain all files for the run directory except the namelist."""
+    asset_list = get_initial_conditions_asset_list(config)
+    asset_list += get_base_forcing_asset_list(config)
+    asset_list += get_orographic_forcing_asset_list(config)
+    asset_list.append(get_field_table_asset(config))
+    asset_list.append(get_diag_table_asset(config))
+    asset_list.append(get_data_table_asset(config))
+    if 'patch_files' in config:
+        if is_dict_or_list(config['patch_files']):
+            asset_list += ensure_is_list(config['patch_files'])
+        else:
+            raise ConfigError('patch_files item in config dictionary must be a dict or list')
+    return asset_list
+
+
+def link_file(source_item, target_item):
+    if os.path.exists(target_item):
+        os.remove(target_item)
+    os.link(source_item, target_item)
+
+
+def copy_file(source_path, target_path):
+    if is_gsbucket_url(source_path):
+        if gsutil_is_installed():
+            check_call(['gsutil', 'cp', source_path, target_path])
+        else:
+            logging.warning(f'Optional dependency gsutil not found. File {source_path} will not be copied to {target_path}')
```

Comment:
Please raise an exception instead of a warning. I think this counts as a catastrophic failure.

Reply:
Done. Also now raise the exception before trying to generate an asset_list from a gcs url if gsutil is not installed.